### PR TITLE
Create user_org

### PR DESCRIPTION
**Pull request by dbrandt1990**
- Created UserOrganizations join table to link Organizations with Users

Userorg relation

- User_orgs  have the attributes "role" and "approvalStatus" which will be set by the Organization. 
    - The "role" will act as a key to access certain features in the app and/or act on behalf of an organization.
    - The "approvalStatus" will have to be set to "APPROVED" by the Organization for the User to be "active" with the org.
        - There is also an option for "REVOKED" which can be used to manage non-compliant volunteers and avoid re-approving 
           them for future opportunities.

To test relations first create a User, I am using [DBeaver](https://dbeaver.com/docs/wiki/Query-Manager):

insert into users (first_name, last_name, email, password)
values(
	'Danny',
	'Brandt',
	'mymail@email.com',
	'123password'
);

usermade

Our User in the DB
user persist

Then create an Organizations:

insert into organizations (name, description, website, phone, city, state, ein, tax_exempt_id, address)
values(
    'Very Good Donators',
    'We donate things',
    'www.gooddonators.com',
    '555-555-5555',
    'Newark',
    'New Jersey',
    123456789,
    987654321,
    '123 Olde Springs Rd'
);

org made

Organization in DB
org persist

Finally create a User_org using your users id and your organizations id:

insert into "userOrganizations" (user_id, org_id, "approvalStatus" , role)
values(
	5,  //user id
	1,  // organization id
	'APPROVED',
	'ADMIN'
);

user_org made

User_org in DB, You can check the relationship by clicking the box/arrow next to the id, and it will pull up the associated User or Organization

Screenshot 2021-12-23 120639

Using Postman we can test the controller routes, you will need to comment out the  @UseGuards(JwtAuthGuard) to avoid getting an unauthorized status code.
POST: 
Request from http://localhost:3001/api/userOrganizations use JSON below for BODY (make sure you use a user_id that YOU have in YOUR DB)
{
    "user_id": 5,
    "org_id":1
}

POST

We can see the  values of approvalStatus  and role default to PENDING and ADMIN

GET :

To get ALL, request from http://localhost:3001/api/userOrganizations

GET all

Find by id using http://localhost:3001/api/userOrganizations/1

GET with id

UPDATE:
In Postman, set the URL request to PATCH and copy JSON below, paste in the BODY tab and send request.
 {
    "approvalStatus":"APPROVED",
    "role":"ADMIN"
}

Before PATCH

PRE PATCH user org

After PATCH

PATCHED userorg

DELETE:

Using http://localhost:3001/api/userOrganizations/1 with DELETE selected

DELETED userorg

We see the typeorm DeleteResults being returned after the DELETE.

Resources:
[DBeaver](https://dbeaver.com/docs/wiki/)
[Postman](https://www.postman.com/downloads/)
[Many-to-many](https://typeorm.io/#/many-to-many-relations)

### CODE REVIEW

**Comment by esteban-gs** [comment]
@dbrandt1990 , there are merge conflicts. Also your branch has diverged from `main`. Can you rebase to main to merge main into your branch?

File: server/src/auth/auth.service.ts [hunk 1]
```diff
@@ -1,14 +1,20 @@
...
 import { User } from '../users/entities/user.entity';
+<<<<<<< HEAD
+<<<<<<< HEAD
+import { jwtConstants } from './constants';
+=======
+import { JwtService } from '@nestjs/jwt';
+>>>>>>> updated import paths in user and auth service
+=======
 import { jwtConstants } from './constants';
+>>>>>>> merge confilct fix
```
Comment (esteban-gs) [inline server/src/auth/auth.service.ts:13]: Can you remove these, we missed some when rebasing

File: server/src/organizations/entities/organization.entity.ts [hunk 2]
```diff
@@ -31,4 +32,7 @@ export class Organization {
 
   @Column('int')
   tax_exempt_id: number;
+  // added for many to many relationship with user_org
```
Comment (esteban-gs) [inline server/src/organizations/entities/organization.entity.ts:35]: You can remove this comment

**Comment by esteban-gs** [comment]
@dbrandt1990 , should I expect the new controllers routes to work for this PR?

File: server/src/auth/auth.service.ts (hunk 1 above)
Comment (dbrandt1990) [reply server/src/auth/auth.service.ts]: This looks like jwtConstants is actually in use, I can't run the server without it.

**Comment by dbrandt1990** [comment]
Yes controller routes should be working.

File: server/src/auth/auth.service.ts (hunk 1 above)
Comment (jd2rogers2) [reply server/src/auth/auth.service.ts]: he meant remove the `>>>>>`, `=======`, and `<<<<<<` lines
these are the merge conflict lines
lmk if you need help resolving

**Comment by jd2rogers2** [comment]
your diff looks like your formatter is making changes to a bunch of files
this makes it hard to parse your changes from the format changes
any way we can make this PR only have changes regarding the user_org join table?

**Comment by esteban-gs** [comment]
@dbrandt1990 , these folders and files should be kabob-case:

```
src/user_org
src/user_org/dto/create-user_org.dto.ts
src/user_org/dto/update-user_org.dto.ts
src/user_org/entities/user_org.entitiy.ts
src/user_org/user_org.controller.ts
src/user_org/user_org.module.ts
src/user_org/user_org.service.ts
```

**Comment by esteban-gs** [comment]
Looks like it's ready on my side @jd2rogers2 , any other concerns?

File: server/src/user-org/entities/user-org.entitiy.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+@Entity('userOrganizations')
+export class UserOrganization {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column('int')
+  user_id!: number;
+
+  @Column('int')
+  org_id!: number;
```
Comment (jd2rogers2) [inline server/src/user-org/entities/user-org.entitiy.ts:33]: what are these `!`s?

File: server/src/user-org/user-org.controller.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
+import type { CreateUserOrganizationDto } from './dto/create-user-org.dto';
+import type { UpdateUserOrganizationDto } from './dto/update-user-org.dto';
+import { UserOrganization } from './entities/user-org.entitiy';
+import { UserOrganizationsService } from './user-org.service';
+
+@Controller('userOrganizations')
+export class UserOrganizationsController {
+  constructor(private readonly userOrganizationsService: UserOrganizationsService) {}
+  @Post()
```
Comment (jd2rogers2) [inline server/src/user-org/user-org.controller.ts:24]: can we add an empty line above `@Post()` and below `constructor?

File: server/src/user-org/user-org.controller.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @UseGuards(JwtAuthGuard)
+  @Get()
+  async findAll() {
+    const allUserOrgs = await this.userOrganizationsService.findAll();
+    return allUserOrgs;
+  }
+
+  @UseGuards(JwtAuthGuard)
```
Comment (jd2rogers2) [inline server/src/user-org/user-org.controller.ts:46]: we no longer have a `JwtAuthGuard` use `CookieGuard` from https://github.com/Nonprofit-Exchange-Hub/web-app/pull/91

File: server/src/user-org/user-org.controller.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      throw new HttpException(
+        { status: HttpStatus.CONFLICT, message: `${error}` },
+        HttpStatus.CONFLICT,
+      );
+    }
+  }
+
+  @UseGuards(JwtAuthGuard)
+  @Get()
+  async findAll() {
```
Comment (jd2rogers2) [inline server/src/user-org/user-org.controller.ts:41]: `async findAll(): Promise<UserOrganization[]> {`

File: server/src/user-org/user-org.service.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    return userOrg;
+  }
+
+  //TODO: Assess if there is a better way than making two requests.
+  async update(id: number, updateUserOrganizationsDto: UpdateUserOrganizationDto) {
+    await this.userOrganizationsRepository.update(id, updateUserOrganizationsDto);
+    return this.userOrganizationsRepository.findOne(id);
+  }
+
+  async remove(id: number) {
```
Comment (jd2rogers2) [inline server/src/user-org/user-org.service.ts:36]: add return types to all of these functions, should be pretty much the same as the return types in the controller

File: server/src/user-org/user-org.service.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import type { Repository } from 'typeorm';
+import { CreateUserOrganizationDto } from './dto/create-user-org.dto';
+import { UpdateUserOrganizationDto } from './dto/update-user-org.dto';
+import { UserOrganization } from './entities/user-org.entitiy';
+
+//only changed the names here still need to change logic
```
Comment (jd2rogers2) [inline server/src/user-org/user-org.service.ts:8]: you can probs remove this comment now right?

**Review by jd2rogers2 — COMMENTED** [review]
couple nits and some merge conflicts that you'll need to fix
but let's also try to sort this issue https://github.com/Nonprofit-Exchange-Hub/web-app/issues/126 out before we land this
but besides those ☝️ this is looking ready, should be a quick approve on next round

**Comment by jd2rogers2** [comment]
```
jamesrogers@jamesrogpremiseMBP-2 web-app % curl -X DELETE localhost:3001/api/userOrganizations/4
{"raw":[],"affected":1}%                                                                     
jamesrogers@jamesrogpremiseMBP-2 web-app % curl -X POST localhost:3001/api/userOrganizations \
-H "Content-Type: application/json" \
-d '{ "approvalStatus": "PENDING","role": "ADMIN", "user": { "id": "18" }, "organization": { "id": "1" } }'
{"approvalStatus":"PENDING","role":"ADMIN","user":{"id":"18"},"organization":{"id":"1"},"id":5,"created_at":"2022-01-29T23:45:13.837Z"}%                                                  
jamesrogers@jamesrogpremiseMBP-2 web-app % curl localhost:3001/api/userOrganizations
[{"id":5,"approvalStatus":"PENDING","role":"ADMIN","created_at":"2022-01-29T23:45:13.837Z","organization":{"id":1,"name":"name","description":"description","website":"website.com","address":"address","phone":"phone","city":"seattle","state":"wa","ein":1,"tax_exempt_id":1},"user":{"id":18,"first_name":"test","last_name":"user","email":"test@test.com"}}]% 
jamesrogers@jamesrogpremiseMBP-2 web-app % curl localhost:3001/api/users/18
{"id":18,"first_name":"test","last_name":"user","email":"test@test.com"}%   
```

☝️ testing for user orgs and testing removing pw from user gets
👇 testing including the user with org get (to prove relationship, minor tweaking needed but we'll figure out which way the `eager`s should be later)

Screen Shot 2022-01-29 at 3 50 17 PM

**Comment by jd2rogers2** [comment]
more resources on user pw
https://stackoverflow.com/questions/50360101/how-to-exclude-entity-field-from-returned-by-controller-json-nestjs-typeorm
https://typeorm.io/#/select-query-builder/hidden-columns

File: server/src/user-org/dto/create-user-org.dto.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { Organization } from '../../organizations/entities/organization.entity';
+import { User } from '../../users/entities/user.entity';
+import { ApprovalStatus, Role } from '../constants';
+
+export class CreateUserOrganizationDto {
+  approvalStatus: ApprovalStatus;
+  organization: Organization;
+  role: Role;
+  user: User;
```
Comment (esteban-gs) [inline server/src/user-org/dto/create-user-org.dto.ts:9]: Are any of these required? we could use the dto validator library

File: server/src/user-org/user-org.controller.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    } catch (error) {
+      throw new HttpException(
+        { status: HttpStatus.CONFLICT, message: `${error}` },
+        HttpStatus.CONFLICT,
+      );
+    }
+  }
+
+  @Delete(':id')
+  async remove(@Param('id') id: string): Promise<DeleteResult> {
```
Comment (esteban-gs) [inline server/src/user-org/user-org.controller.ts:76]: For all the controller methods using an id that should be an int use the `ParseIntPipe`, and follow the example in [NestJs](https://docs.nestjs.com/pipes)

File: server/src/users/users.controller.ts [hunk 11]
```diff
@@ -18,7 +18,7 @@ export class UsersController {
   }
 
   @Get(':id')
-  findOne(@Param('id') id: string) {
+  async findOne(@Param('id') id: string) {
```
Comment (esteban-gs) [inline server/src/users/users.controller.ts:21]: #### Fix 
Ignore if id is in fact a string

For all the controller methods using an id that should be an int use the `ParseIntPipe`, and follow the example in [NestJs](https://docs.nestjs.com/pipes)

File: server/src/user-org/user-org.controller.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @Get()
+  async findAll(): Promise<UserOrganization[]> {
+    const allUserOrgs = await this.userOrganizationsService.findAll();
+    return allUserOrgs;
+  }
+
+  @Get(':id')
+  async findOne(@Param('id') id: string): Promise<UserOrganization> {
```
Comment (esteban-gs) [inline server/src/user-org/user-org.controller.ts:45]: #### Fix 
Ignore if id is in fact a string

For all the controller methods using an id that should be an int use the `ParseIntPipe`, and follow the example in [NestJs](https://docs.nestjs.com/pipes)

File: server/src/user-org/user-org.controller.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        { staus: HttpStatus.NOT_FOUND, message: 'User relation to Organization not found' },
+        HttpStatus.NOT_FOUND,
+      );
+    }
+    return userOrg;
+  }
+
+  @Patch(':id')
+  async update(
+    @Param('id') id: string,
```
Comment (esteban-gs) [inline server/src/user-org/user-org.controller.ts:58]: #### Fix 
Ignore if id is in fact a string

For all the controller methods using an id that should be an int use the `ParseIntPipe`, and follow the example in [NestJs](https://docs.nestjs.com/pipes)

**Review by esteban-gs — CHANGES_REQUESTED** [review]

File: server/src/user-org/dto/create-user-org.dto.ts (hunk 9 above)
Comment (jd2rogers2) [reply server/src/user-org/dto/create-user-org.dto.ts:9]: good call, added some of the validators that we use elsewhere, where appropriate
these are required by default right?

**Review by esteban-gs — APPROVED** [review]
